src/mono/mono/mini/jit-icalls.c

```diff
@@ -1482,7 +1483,9 @@ mono_gsharedvt_constrained_call (gpointer mp, MonoMethod *cmethod, MonoClass *kl
 		g_assert (fsig->param_count < 16);
 		memcpy (new_args, args, fsig->param_count * sizeof (gpointer));
 		for (int i = 0; i < fsig->param_count; ++i) {
-			if (deref_args [i])
+			// If the argument is a ref or nullable non-vtype, deref it
```

Comment:
@vargaz Please confirm this is correct.

Reply:
Isn't this information already known in the wrapper and passed through `deref_args` ? Also isn't `deref_args` supposed to be set only for valuetype args, why the check below ?

Reply:
The case we want handle here is `handler.Test<bool?>(true);`.

> Isn't this information already known at the callsite in `handle_constrained_gsharedvt_call` and can be passed through `deref_args`?

Yes, this information already known at the callsite:
```csharp
if (mini_is_gsharedvt_type (fsig->params [i])) {
	MonoInst *is_deref;
	int deref_arg_reg;
	ins = mini_emit_get_gsharedvt_info_klass (cfg, mono_class_from_mono_type_internal (fsig->params [i]), MONO_RGCTX_INFO_CLASS_BOX_TYPE);
	deref_arg_reg = alloc_preg (cfg);
	/* deref_arg = BOX_TYPE != MONO_GSHAREDVT_BOX_TYPE_VTYPE */
	EMIT_NEW_BIALU_IMM (cfg, is_deref, OP_ISUB_IMM, deref_arg_reg, ins->dreg, 1);
	MONO_EMIT_NEW_STORE_MEMBASE (cfg, OP_STOREI1_MEMBASE_REG, is_gsharedvt_ins->dreg, i, is_deref->dreg);
}
```
> Also isn't `deref_args` supposed to be set only for valuetype args, why the check below?

It appears that an argument is being dereferenced if the `BOX_TYPE != MONO_GSHAREDVT_BOX_TYPE_VTYPE`. However, if the argument is a nullable value type, `deref_args [i]` is set because `mono_class_is_nullable` is checked first:
```csharp
if (MONO_TYPE_IS_REFERENCE (m_class_get_byval_arg (impl_class)))
	return GUINT_TO_POINTER (MONO_GSHAREDVT_BOX_TYPE_REF);
else if (mono_class_is_nullable (impl_class))
	return GUINT_TO_POINTER (MONO_GSHAREDVT_BOX_TYPE_NULLABLE);
else
	return GUINT_TO_POINTER (MONO_GSHAREDVT_BOX_TYPE_VTYPE);
```
Then, during runtime, a nullable value type is dereferenced, leading to a runtime crash. Perhaps this check (if not nullable vtype) can be moved to the `handle_constrained_gsharedvt_call`, which might result in larger codegen.


> Is there a testcase which fails for this case ?

The fullAOT tests are not currently running on the CI. Once the job is enabled, we will check if this case is already covered. If it is not covered, we will add a runtime test. You can currently verify this by using the following sample: https://github.com/dotnet/runtime/issues/94467#issuecomment-1801332061.

Reply:
Would suggest this fix instead:
```
diff --git a/src/mono/mono/mini/jit-icalls.c b/src/mono/mono/mini/jit-icalls.c
index 21f3252c814..442cb9a65aa 100644
--- a/src/mono/mono/mini/jit-icalls.c
+++ b/src/mono/mono/mini/jit-icalls.c
@@ -1482,7 +1482,7 @@ mono_gsharedvt_constrained_call (gpointer mp, MonoMethod *cmethod, MonoClass *kl
 		g_assert (fsig->param_count < 16);
 		memcpy (new_args, args, fsig->param_count * sizeof (gpointer));
 		for (int i = 0; i < fsig->param_count; ++i) {
-			if (deref_args [i])
+			if (deref_args [i] != MONO_GSHAREDVT_BOX_TYPE_VTYPE && deref_args [i] != MONO_GSHAREDVT_BOX_TYPE_NULLABLE)
 				new_args [i] = *(gpointer*)new_args [i];
 		}
 		args = new_args;
diff --git a/src/mono/mono/mini/method-to-ir.c b/src/mono/mono/mini/method-to-ir.c
index 99a11126b62..1f359e31cc6 100644
--- a/src/mono/mono/mini/method-to-ir.c
+++ b/src/mono/mono/mini/method-to-ir.c
@@ -3896,13 +3896,8 @@ handle_constrained_gsharedvt_call (MonoCompile *cfg, MonoMethod *cmethod, MonoMe
 			int addr_reg;
 
 			if (mini_is_gsharedvt_type (fsig->params [i])) {
-				MonoInst *is_deref;
-				int deref_arg_reg;
 				ins = mini_emit_get_gsharedvt_info_klass (cfg, mono_class_from_mono_type_internal (fsig->params [i]), MONO_RGCTX_INFO_CLASS_BOX_TYPE);
-				deref_arg_reg = alloc_preg (cfg);
-				/* deref_arg = BOX_TYPE != MONO_GSHAREDVT_BOX_TYPE_VTYPE */
-				EMIT_NEW_BIALU_IMM (cfg, is_deref, OP_ISUB_IMM, deref_arg_reg, ins->dreg, 1);
-				MONO_EMIT_NEW_STORE_MEMBASE (cfg, OP_STOREI1_MEMBASE_REG, is_gsharedvt_ins->dreg, i, is_deref->dreg);
+				MONO_EMIT_NEW_STORE_MEMBASE (cfg, OP_STOREI1_MEMBASE_REG, is_gsharedvt_ins->dreg, i, ins->dreg);
 			} else if (has_gsharedvt) {
 				MONO_EMIT_NEW_STORE_MEMBASE_IMM (cfg, OP_STOREI1_MEMBASE_IMM, is_gsharedvt_ins->dreg, i, 0);
 			}
```